test/unit/app/controllers/swaps-test.js

```diff
@@ -865,37 +1278,101 @@ describe('SwapsController', function () {
   })
 
   describe('utils', function () {
-    describe('getMedian', function () {
-      const { getMedian } = utils
+    describe('getMedianEthValueQuote', function () {
+      const { getMedianEthValueQuote } = utils
 
       it('calculates median correctly with uneven sample', function () {
-        const values = [3, 2, 6].map((value) => new BigNumber(value))
-        const median = getMedian(values)
+        const expectedResult = {
+          ethFee: '10',
+          metaMaskFeeInEth: '5',
+          ethValueOfTokens: '0.3',
+        }
+        const values = [
+          {
+            overallValueOfQuote: '3',
+            ethFee: '10',
+            metaMaskFeeInEth: '5',
+            ethValueOfTokens: '0.3',
+          },
+          {
+            overallValueOfQuote: '2',
+            ethFee: '20',
+            metaMaskFeeInEth: '3',
+            ethValueOfTokens: '0.2',
+          },
+          {
+            overallValueOfQuote: '6',
+            ethFee: '40',
+            metaMaskFeeInEth: '6',
+            ethValueOfTokens: '0.6',
+          },
+        ]
 
-        assert.strictEqual(
-          median.toNumber(),
-          3,
-          'should have returned correct median',
+        const median = getMedianEthValueQuote(values)
+
+        assert.deepEqual(
+          median,
+          expectedResult,
+          'should have returned correct median quote object',
         )
       })
 
       it('calculates median correctly with even sample', function () {
```

Comment:
This test is... interesting. I was confused at first about why `ethFee` was `20` given that the median was split between 2 and 3, then I realized that one of the `2` quotes was arbitrarily chosen over the other as the source of data.

I think this shows that the median calculation is broken. We should be considering _all_ quotes with the median value as being representative of the median (or all quotes on either side of a split, as in this case), not just one arbitrary selection from the group.

Reply:
TIL how statistics work apparently. This was an unexpected discover :sweat_smile:

Reply:
oh, wow, very nice catch. The median of the overallValueOfQuote is correctly the average/mean of the middle two numbers, but the other values we calculate are not exactly medians are they... so we should get the mean of all values associated with a overallValueOfQuote that equals either of the middle values used to calculate the median overallValueOfQuote.

This can lead to some interesting edge cases. Imagine 8 quotes, 4 with an overall value of X and 4 with an overall value of Y. The `ethFee`, `metaMaskFeeInEth` and `ethValueOfTokens` values that we associate with the median will just be averages/means of each of those values from each quote.

I think you are right that this is most correct though. And an even number of quote with two identical overall values will be very very rare.

Reply:
Actually it's more complicated than that.

```
[
  {
    overallValueOfQuote: '2',
    ethFee: AX,
    metaMaskFeeInEth: AY,
    ethValueOfTokens: AZ,
  },
  {
    overallValueOfQuote: '2',
    ethFee: BX,
    metaMaskFeeInEth: BY,
    ethValueOfTokens: BZ,
  },
  {
    overallValueOfQuote: '3',
    ethFee: CX,
    metaMaskFeeInEth: CY,
    ethValueOfTokens: CZ,
  },
  {
    overallValueOfQuote: '6',
    ethFee: DX,
    metaMaskFeeInEth: DY,
    ethValueOfTokens: DZ,
  },
]
```

In this case, the median is split between `2` and `3`. But we don't want to sum all three `2` quotes along with the `3` quote and divide by 4, because that would too heavily weight the `2` quotes.

Instead I think we should sum all of the `2` quotes and divide by the number of `2` quotes, and do the same for the `3` quotes. And _then_ we'd sum these results and divide by two, as we're doing now.

Reply:
Thanks @Gudahtt 

I believe the latest commit correctly handles this, in accordance with both of your comments above. Unit tests are added for these cases as well. ec9f935b9